2.4.0

KillerOfPie opened the pull request:
[contributing]: https://github.com/SparklingComet/TradeShop/blob/master/.github/CONTRIBUTING.md

## Pull Request Etiquette

There are several guidelines you should follow in order for your Pull Request to be merged.
Replace `[ ]` with `[x]` to cross the checkbox.

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

> It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.

## Changes

_Replace this sentence with a description of che changes introduced by this PR._

SparklingComet committed:
Restrain power of TSAF (addons cannot override protection config)

SparklingComet committed:
Restrain power of TSAF (addons cannot override protection config)

SparklingComet committed:
Merge remote-tracking branch 'origin/tweak-hopper-protection' into tweak-hopper-protection

KillerOfPie committed:
Permissions and assorted fixes
- Removed duplicate message(NO_EDIT > NO_SHOP_PERMISSION)
- `Switch` command now requires `EDIT` permission and the create permission for the shop it is switching to
- `Edit` command now requires `EDIT` permission as well as proper role on shop or `ADMIN` permission
- `What` command now requires `EDIT` permission; Not sure if I want this to be this way, if we remove this the wiki needs to be updated to reflect it.

KillerOfPie committed:
Shulker Dupe Fix and other various changes
- Version bump to 2.3.2-DEV
- Fixed Errors caused by preventing shulkers as cost/product when storage is a shulker box when no storage was present(mostly iTrade shops)
- Re-organized Commands enum list to match wiki sections
- Changed some variable names to be more specific
- Utils#getItems, fix for Shulker dupe was missing #clone()

KillerOfPie committed:
Overstack trading is back...
- Re-added support for overstack trading(the reasons for removing it were no longer a problem)
- Multistack support is now for items that would not trade the same
- Add(Cost/Product) now add ontop of existing items if they match
- Data will be the same as the first item added, i.e. If you add an item with a custom name then disable custom name checking and add an item without a custom name, the custom name would stay on the edit menu and sign.

SparklingComet commented:
#108 #110

KillerOfPie committed:
Added creation commands and other minor tweaks

KillerOfPie committed:
Updated resource Messages and Settings
- I don't think these are needed anymore, but leaving for reference.

KillerOfPie committed:
2.3.2-STABLE + Setting comment fixes

KillerOfPie committed:
Fix for Sign attached block protection and Setting comment changes

KillerOfPie committed:
Fix for not being able to destroy adjacent signs & new Message

SparklingComet reviewed (commented):
This warrants a bump in version to minor release level, right? As in, directly 2.4.0.

SparklingComet committed:
Merge pull request #107 from Tradeshop/tweak-hopper-protection

Restrain power of TSAF (addons cannot override protection config)

SparklingComet commented:
Changes from #107 are partially untested and need to be tested before we release.

KillerOfPie commented:
> This warrants a bump in version to minor release level, right? As in, directly 2.4.0.

They were fairly minor changes, I see no need to bump minor. But there also no reason we can't.

SparklingComet commented:
> > This warrants a bump in version to minor release level, right? As in, directly 2.4.0.
> 
> They were fairly minor changes, I see no need to bump minor. But there also no reason we can't.

The following (from the latest dev release) warrants a minor version bump (x.Y.z) instead of a patch bump (x.y.Z):

> New Features
> Ability to turn existing signs into Shops using /ts (create/createbi/createi)

KillerOfPie committed:
Fix for EnchantedBooks not being compared properly
- Added compare for EnchantmentStorageMeta MetaType
- Minor tweaks and Variable name fixes

KillerOfPie commented:
`if (((TradeShop) Bukkit.getPluginManager().getPlugin("TradeShop")).isFrozen()) return;`
https://github.com/Tradeshop/TradeShop/blob/7824f0c265fd4883892c9d9fe6a651b778b86b0c/src/main/java/org/shanerx/tradeshop/objects/Shop.java#L733

What is this for? it seems to be constantly stopping the updateSign() method from running?
^ I'm seeing this when I'm trying to create signs using the commands.

also in most places you can either use `plugin` or `utils.getPlugin` instead of `((TradeShop) Bukkit.getPluginManager().getPlugin("TradeShop"))`

> > > This warrants a bump in version to minor release level, right? As in, directly 2.4.0.
> > 
> > 
> > They were fairly minor changes, I see no need to bump minor. But there also no reason we can't.
> 
> The following (from the latest dev release) warrants a minor version bump (x.Y.z) instead of a patch bump (x.y.Z):
> 
> > New Features
> > Ability to turn existing signs into Shops using /ts (create/createbi/createi)

Bumping to 2.4.0 with my next set of changes and also adding in the toggleable Admin mode since that is a small addition and makes me feel better about making it a minor release.

kyrtion commented:
```- `What` command now requires `EDIT` permission; Not sure if I want this to be this way, if we remove this the wiki needs to be updated to reflect it.```
No. `What` command now need requires `INFO` permission. Why? `What` - permission, what this is shop and list products, costs and etc.

KillerOfPie commented:
> `` - `What` command now requires `EDIT` permission; Not sure if I want this to be this way, if we remove this the wiki needs to be updated to reflect it. `` No. `What` command now need requires `INFO` permission. Why? `What` - permission, what this is shop and list products, costs and etc. @KillerOfPie

The commit note was wrong and they can't be edited afaik, if you look at the first 2.3.2 release notes it is correct in that.

KillerOfPie committed:
Added Toggle-able Admin mode  and other fixes
- Change Ver to 2.4.0-DEV
- Added `TOGGLE_ADMIN` command
- Cleaned up Message sending to use Message methods wherever possible
- Added Permissions#isAdminEnabled that checks for internal/external permissions and the users current mode
- Removed un-needed permission check in EditCommand
- Changed Message#sendMessage to use Tuple array for replacements
- Removed Firework Compare options from Config, if re-added all changebale compare options should have a config setting
- Moved updateSign to happen at the end of Shop saving.

SparklingComet commented:
> What is this for? it seems to be constantly stopping the updateSign() method from running?
@KillerOfPie Correct, it is a hacky fix for that very tricky bug that kept us confused for months because of automatic early sign updates.

KillerOfPie committed:
Minor bug Fixes, New Admin Command, and Reformatting
- Moved Commands from enumys to the commands package
- Added ADMIN command to view or directly set a players Admin Mode
- Changed default ADMIN_TOGGLED message from `now` mto `currently`. I think this makes more sense when the command is used both after changing and for viewing
- Changed default value of adminEnabled to true. Still requires ADMIN permissions so having the value true for non admin players wont do anything
- Removed Freeze and replaced with a `s != null` check on sign update. This fixes signs made with commands not updating and I believe still fixes what freeze was added to fix as well.

KillerOfPie committed:
Removed some Freeze usages I had missed

KillerOfPie committed:
Added missing `#` on message description newLine

SparklingComet commented:
Is the new fix tested extensively?

KillerOfPie commented:
> Is the new replacement fix tested extensively?

No, I never got it on my test server so I can't know for dire. but the problem was just an error in console so

KillerOfPie committed:
Change to SHOP_INSUFFICIENT_ITEMS message and Shop trade feedback messages

KillerOfPie committed:
More Unification of message sending

KillerOfPie committed:
Fixed Attachment Protection protecting none Attached blocks

KillerOfPie committed:
Bug fixes and minor changes
- Added ITRADE_NO_COST_TEXT setting for when iTrade shops trade without a Cost, replaces the {AMOUNT} {NAME}
- Added Setting warning to not exceed 6 Multitrade as it can potentially cause errors with iTrade shops. Not blocking outright as it probably wont be a problem unless the resulting amount of items take up more the 54 slots...which i guess wouldn't fit on the player anyways...
- Added getCostItemStacks and getProductItemStacks to shop returning ItemStacks instead of ShopItemStacks
- Changed getCost and getProduct to return clones instead of the actual objects to prevent accidental changes and fix iTrade interstack bug
- Cleaned up if statement repetition in ShopProtectionListener
- Fixed trade multiplier not working for iTrade shops

KillerOfPie committed:
Bug fix and bad debug removal
- Removed temporary debug code that was missed
- Fix NPE error from clone in ShopItemStack

KillerOfPie committed:
Bug fix and other minor changes
- Fix NPE from ShopItemStack#getAmount
- Removed unneeded getItemStack calls for ^

SparklingComet committed:
Set Version tag to STABLE

SparklingComet commented:
Merging and releasing.
